dadi/frontend/lib/keyboard.js

```diff
@@ -0,0 +1,176 @@
+      'down': 40,
+      'insert': 45,
+      'delete': 46,
+      'command': 91,
+      'left command': 91,
+      'right command': 93,
+      'numpad *': 106,
+      'numpad +': 107,
+      'numpad -': 109,
+      'numpad .': 110,
+      'numpad /': 111,
+      'num lock': 144,
+      'scroll lock': 145,
+      'my computer': 182,
+      'my calculator': 183,
+      ';': 186,
+      '=': 187,
+      ',': 188,
+      '-': 189,
+      '.': 190,
+      '/': 191,
+      '`': 192,
+      '[': 219,
+      '\\': 220,
+      ']': 221,
+      "'": 222,
+      // Alias
+      'windows': 91,
+      '⇧': 16,
+      '⌥': 18,
+      '⌃': 17,
+      '⌘': 91,
+      'ctl': 17,
+      'control': 17,
+      'option': 18,
+      'pause': 19,
+      'break': 19,
+      'caps': 20,
+      'return': 13,
+      'escape': 27,
+      'spc': 32,
+      'pgup': 33,
+      'pgdn': 34,
+      'ins': 45,
+      'del': 46,
+      'cmd': 91
+    }
+    // Letters
+    for (i = 97; i < 123; i++) codes[String.fromCharCode(i)] = i - 32
+    // numbers
+    for (var i = 48; i < 58; i++) codes[i - 48] = i
+    // Function keys
+    for (i = 1; i < 13; i++) codes['f'+i] = i + 111
+
+    // Numpad keys
+    for (i = 0; i < 10; i++) codes['numpad '+i] = i + 96
+
+    return codes
+  }
+
+  find (pattern) {
+    return pattern.map(key => {
+      return this.keys[key]
+    }).filter(key => {
+      return key
+    })
+  }
+}
+
+class Pattern {
+  constructor (pattern, keys) {
+    this.pattern = pattern
+    this.keys = keys
+    this.active = 0
+  }
+
+  next (keyCode) {
+    if (this.match(keyCode, this.keys[this.active])) {
+      if (this.active < this.keys.length -1) {
+        this.active ++
+        return true
+      } else {
+        this.callback({
+          pattern: this.pattern,
+          keys: this.keys
+        })
+        this.reset()
+        return true
+      }
+    } else {
+      this.reset()
+      return false
+    }
+  }
+  match (key, current) {
+    return key === current
+  }
+
+  reset () {
+    this.active = 0
+  }
+
+  do (callback) {
+    this.callback = callback
+  }
+}
+
+export class Keyboard extends Keys {
+  constructor () {
+    super()
+    this.shortcuts = []
+    this.listen()
+  }
+
+  listen () {
+    if (window) {
+      window.addEventListener('keydown', this.keydown.bind(this))
+      window.addEventListener('keyup', this.keyup.bind(this))
+    }
+  }
+
+  keydown (event) {
+    if (event.keyCode) {
+      if (this.findShortcut(event.keyCode)) {
+        event.preventDefault()
+      }
+    }
+  }
+
+  keyup (event) {
+    this.shortcuts.forEach(pattern => {
+      pattern.reset()
+    })
+    event.preventDefault()
+  }
+
+  findShortcut (keyCode) {
+    return this.shortcuts.find(shortcut => {
+      return shortcut.next(keyCode)
+    })
+  }
+
+  on (pattern) {
+    let keys = this.find(pattern.toLowerCase().split('+'))
+    let shortcut = new Pattern(pattern, keys)
+    this.shortcuts.push(shortcut)
+    return shortcut
```

Comment:
It'd be great to have more consistency around spacing (perhaps something we could add to the doc). I think having a blank line after the declarations bit of every block (so after the two `let`) and before the `return` would make things more readable. You've done that in some parts of the PR.